tony-history-server/app/utils/ParserUtils.java

```diff
@@ -30,28 +32,52 @@
 public class ParserUtils {
   private static final Logger.ALogger LOG = Logger.of(ParserUtils.class);
 
-  public static JobMetadata parseMetadata(FileSystem fs, Path metadataFilePath) {
-    if (!pathExists(fs, metadataFilePath)) {
-      return new JobMetadata();
+  @VisibleForTesting
+  static boolean isValidMetadata(String[] metadata, String jobIdRegex) {
+    if (metadata.length != 5) {
+      LOG.error("Missing fields in metadata");
+      return false;
     }
+    return metadata[0].matches(jobIdRegex) && metadata[1].matches("\\d+") && metadata[2].matches("\\d+")
+        && metadata[3].matches(metadata[3].toLowerCase()) && metadata[4].equals(metadata[4].toUpperCase());
+  }
 
-    String fileContent = contentOfHdfsFile(fs, metadataFilePath);
+  public static JobMetadata parseMetadata(FileSystem fs, Path jobFolderPath, String jobIdRegex) {
     JobMetadata jobMetadata = new JobMetadata();
-    JsonNode jObj;
+    DateFormat simple = new SimpleDateFormat("dd MMM yyyy HH:mm:ss:SSS Z");
+    String[] metadata;
+
+    if (!pathExists(fs, jobFolderPath)) {
+      return jobMetadata;
+    }
+
     try {
-      jObj = Json.parse(fileContent);
-    } catch (JsonSyntaxException e) {
-      LOG.error("Couldn't parse metadata", e);
+      metadata = Arrays.stream(fs.listStatus(jobFolderPath))
+          .filter(f -> f.getPath().toString().contains("jhist")) // ignore config.xml
+          .map(histFilePath -> HdfsUtils.getJobId(histFilePath.toString()))
+          .map(histFile -> histFile.substring(0, histFile.lastIndexOf('.'))) // remove .jhist
```

Comment:
This chain of stream operations seems weird. You know the jhist file is going to be named `<appId>...jhist` so just directly look for that file. There should also be exactly 1 jhist file, so you could assert that, too. (The code currently still works if there are multiple file matches but that could lead to unexpected behavior later when you try to parse the metadata from the file name.)

Reply:
I don't think we can directly look for that file since the file name depends on other variables (`appId-started-completed-user-status.jhist`) so we can directly look for the file. Also, I was assuming there's just 1 jhist file in the `jobFolderPath`, maybe I should put that in the Javadoc for this function.

Reply:
What I meant was: You can use listStatus with a glob, e.g. `fs.listStatus("<appId>*.jhist")` and then you can check that the returned FileStatus array has length 1.

Reply:
For checking the list is exactly size 1, a common pattern is using Guava's Preconditions: `Preconditions.checkArgument(files.size() == 1)`

Reply:
Oh, that's cool, I didn't know `fs.listStatus` could consume regex.

Reply:
Discussed offline, `fs.listStatus` uses `PathFilter` (no glob) which makes the code more complex than it should be. Fixed: https://github.com/linkedin/TonY/pull/81/commits/ed06cf126ca618f794531e647344310f689b6969.